clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy

```diff
@@ -0,0 +1,265 @@
+      UpsertAmazonLoadBalancerV2Description.TargetGroup targetGroup = targetGroups.find({ it.name == awsTargetGroup.getTargetGroupName() })
+      loadBalancing.modifyTargetGroup(new ModifyTargetGroupRequest()
+        .withTargetGroupArn(awsTargetGroup.targetGroupArn)
+        .withHealthCheckIntervalSeconds(targetGroup.healthCheckInterval)
+        .withHealthCheckPath(targetGroup.healthCheckPath)
+        .withHealthCheckPort(targetGroup.healthCheckPort)
+        .withHealthCheckProtocol(targetGroup.healthCheckProtocol)
+        .withHealthCheckTimeoutSeconds(targetGroup.healthCheckTimeout)
+        .withUnhealthyThresholdCount(targetGroup.unhealthyThreshold)
+        .withMatcher(new Matcher().withHttpCode(targetGroup.healthCheckMatcher))
+      )
+      task.updateStatus BASE_PHASE, "Target group updated in ${loadBalancerName} (${awsTargetGroup.targetGroupName}:${awsTargetGroup.port}:${awsTargetGroup.protocol})."
+
+      // Update attributes
+      String exceptionMessage = modifyTargetGroupAttributes(loadBalancing, awsTargetGroup, targetGroup.attributes)
+      if (exceptionMessage) {
+        amazonErrors << exceptionMessage
+      }
+    }
+
+
+    // Create new target groups
+    def targetGroupsToCreate = targetGroups.findAll { targetGroup ->
+      (existingTargetGroups.find { targetGroup.name == it.targetGroupName &&
+                                    targetGroup.port == it.port &&
+                                    targetGroup.protocol.toString() == it.protocol }) == null
+    }
+    targetGroupsToCreate.each { targetGroup ->
+      TargetGroup createdTargetGroup
+      try {
+        CreateTargetGroupResult createTargetGroupResult = loadBalancing.createTargetGroup(new CreateTargetGroupRequest()
+          .withProtocol(targetGroup.protocol)
+          .withPort(targetGroup.port)
+          .withName(targetGroup.name)
+          .withVpcId(loadBalancer.vpcId)
+          .withHealthCheckIntervalSeconds(targetGroup.healthCheckInterval)
+          .withHealthCheckPath(targetGroup.healthCheckPath)
+          .withHealthCheckPort(targetGroup.healthCheckPort)
+          .withHealthCheckProtocol(targetGroup.healthCheckProtocol)
+          .withHealthCheckTimeoutSeconds(targetGroup.healthCheckTimeout)
+          .withHealthyThresholdCount(targetGroup.healthyThreshold)
+          .withUnhealthyThresholdCount(targetGroup.unhealthyThreshold)
+          .withMatcher(new Matcher().withHttpCode(targetGroup.healthCheckMatcher))
+        )
+        task.updateStatus BASE_PHASE, "Target group created in ${loadBalancerName} (${targetGroup.name}:${targetGroup.port}:${targetGroup.protocol})."
+        createdTargetGroup = createTargetGroupResult.getTargetGroups().get(0)
+      } catch (AmazonServiceException e) {
+        def exceptionMessage = "Failed to create target group ${targetGroup.name} for ${loadBalancerName} - reason: ${e.errorMessage}."
+        task.updateStatus BASE_PHASE, exceptionMessage
+        amazonErrors << exceptionMessage
```

Comment:
Added rollback for removed listeners to this as well. Was a little more work since listener actions are by ARN and we don't pass the ARN around, I had to recreate the listener objects.